r3bbase/R3BTprevTnext.cxx

```diff
@@ -0,0 +1,157 @@
+R3BTprevTnext::R3BTprevTnext()
+    : R3BTprevTnext("R3B Tprev/Tnext", 1)
+{
+}
+
+R3BTprevTnext::R3BTprevTnext(const TString& name, Int_t iVerbose)
+    : FairTask(name, iVerbose)
+{
+}
+
+InitStatus R3BTprevTnext::Init()
+{
+
+    FairRootManager* rootManager = FairRootManager::Instance();
+    if (rootManager == nullptr)
+    {
+        return kFATAL;
+    }
+
+    FairRuntimeDb* rtdb = FairRuntimeDb::instance();
+    if (rtdb == nullptr)
+    {
+        LOG(error) << "FairRuntimeDb not opened!";
+        return kFATAL;
+    }
+
+    fR3BEventHeader = dynamic_cast<R3BEventHeader*>(rootManager->GetObject("EventHeader."));
+    if (fR3BEventHeader == nullptr)
+    {
+        LOG(error) << "R3BTprevTnext::Init() EventHeader. not found";
+        return kFATAL;
+    }
+
+    fSamplerMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMapped"));
+    if (fSamplerMapped == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fSamplerMSMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMSMapped"));
+    if (fSamplerMSMapped == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fMSOffsetPar = dynamic_cast<R3BMSOffsetPar*>(rtdb->getContainer("MSOffsetPar"));
+    if (fMSOffsetPar == nullptr)
+    {
+        LOG(error) << "Could not find MSOffsetPar container!";
+        return kFATAL;
+    }
+
+    return kSUCCESS;
+}
+
+void R3BTprevTnext::Exec(Option_t* /*opt*/)
+{
+    Double_t fTprev = -10.0; // Initialization with invalid values.
+    Double_t fTnext = -10.0;
+    Int_t dMScounter = 0;
+    const Int_t sampHits = fSamplerMapped->GetEntriesFast();
+    const Int_t sampmsHits = fSamplerMSMapped->GetEntriesFast();
+    const Double_t MSOffset = fMSOffsetPar->GetMSOffset();
+    Double_t SAMPTime = 0;
+    Double_t SAMPMSTime = 0;
+    R3BSamplerMappedData* SAMPMapped = nullptr;
+    R3BSamplerMappedData* SAMPMSMapped = nullptr;
+    bool sawMS = false;
+    if (fR3BEventHeader->GetTpat() < fTpat && fR3BEventHeader->GetTrigger() == 1 && fR3BEventHeader->GetTpat() > 0)
+    {
+        if (sampmsHits == 1 && sampHits > 0)
+        {
+            SAMPMSMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMSMapped->At(0));
+            SAMPMSTime = SAMPMSMapped->GetTime();
+            for (Int_t i = 0; i < sampHits; ++i)
+            {
+                SAMPMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMapped->At(i));
+                SAMPTime = SAMPMapped->GetTime();
+                auto tpn = SAMPTime - SAMPMSTime - MSOffset;
+                if (tpn < -fDelta_clk)
+                {
+                    fTprev = -CLOCK_PERIOD * tpn;
+                }
+                if (abs(tpn) <= 1)
+                {
+                    sawMS = true;
+                    ++dMScounter;
+                }
+                if (tpn > fDelta_clk)
+                {
+                    fTnext = CLOCK_PERIOD * tpn;
+                    break;
+                }
+            }
+        }
+    }
+    if (sawMS)
```

Comment:
I am still not happy with the interpretation of sawMS, fTprev and fTnext here. 
* If sawMS is false, we should probably set Tnext and Tprev to some negative header constant INVALID_NOMS, it it keeps happening, we should tell the user that their offset is likely wrong. 
* Otherwise, if fTprev is invalid, we could set it to an estimate on how long our DAQ dead time is minimally. (The fancier way to track samplers over multiple readouts can be implemented later.) Perhaps make it a header constant on the order of 2000ns?
* If fTnext is invalid, the defensible value is the offset, which is the delay through the trigger matrix. (If another LOS trigger had appeared before the MS arrived at the sampler, it would have been recorded, after all.)

43.1e9 aka "no triggers for 43 seconds" should not be used as an output ever.